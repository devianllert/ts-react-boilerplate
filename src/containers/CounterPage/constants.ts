export const COUNTER_INCREMENT = 'boilerplate/Counter/COUNTER_INCREMENT';
export const COUNTER_DECREMENT = 'boilerplate/Counter/COUNTER_DECREMENT';
export const COUNTER_CLEAR = 'boilerplate/Counter/COUNTER_CLEAR';
