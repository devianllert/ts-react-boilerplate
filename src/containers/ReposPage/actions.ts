import {
  ReposRequest,
  ReposSuccess,
  ReposFailure,
  ReposFulfill,
  ReposTrigger,
} from './types';
import {
  REPOS_REQUEST,
  REPOS_FAILURE,
  REPOS_SUCCESS,
  REPOS_FULFILL,
  REPOS_TRIGGER,
} from './constants';
import { Repo } from '../../services/repos.service';

const reposTrigger = (): ReposTrigger => ({
  type: REPOS_TRIGGER,
});

const reposRequest = (): ReposRequest => ({
  type: REPOS_REQUEST,
});

const reposSuccess = (data: Repo[]): ReposSuccess => ({
  type: REPOS_SUCCESS,
  payload: data,
});

const reposFailure = (error: Error): ReposFailure => ({
  type: REPOS_FAILURE,
  payload: error,
});

const reposFulfill = (): ReposFulfill => ({
  type: REPOS_FULFILL,
});

export {
  reposTrigger,
  reposRequest,
  reposSuccess,
  reposFailure,
  reposFulfill,
};
