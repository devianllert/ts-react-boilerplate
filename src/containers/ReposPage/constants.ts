export const REPOS_TRIGGER = 'boilerplate/Repos/REPOS_TRIGGER';
export const REPOS_REQUEST = 'boilerplate/Repos/REPOS_REQUEST';
export const REPOS_SUCCESS = 'boilerplate/Repos/REPOS_SUCCESS';
export const REPOS_FAILURE = 'boilerplate/Repos/REPOS_FAILURE';
export const REPOS_FULFILL = 'boilerplate/Repos/REPOS_FULFILL';
