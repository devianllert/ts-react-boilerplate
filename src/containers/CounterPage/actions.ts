import { CounterIncrement, CounterDecrement, CounterClear } from './types';
import { COUNTER_INCREMENT, COUNTER_DECREMENT, COUNTER_CLEAR } from './constants';

const counterIncrement = (): CounterIncrement => ({
  type: COUNTER_INCREMENT,
});

const counterDecrement = (): CounterDecrement => ({
  type: COUNTER_DECREMENT,
});

const counterClear = (): CounterClear => ({
  type: COUNTER_CLEAR,
});

export {
  counterIncrement,
  counterDecrement,
  counterClear,
};
