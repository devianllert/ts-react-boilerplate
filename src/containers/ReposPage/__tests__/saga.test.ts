import {
  put,
  takeLatest,
  PutEffect,
  CallEffect,
} from 'redux-saga/effects';

import {
  ReposFulfill,
  ReposRequest,
  ReposSuccess,
  ReposFailure,
} from '../types';
import { Repo } from '../../../services/repos.service';

import { REPOS_TRIGGER } from '../constants';
import { reposSuccess, reposFailure } from '../actions';

import reposData, { getRepos } from '../saga';

describe('getRepos Saga', () => {
  let getReposGenerator: Generator<
  | PutEffect<ReposRequest>
  | CallEffect<Repo[]>
  | PutEffect<ReposSuccess>
  | PutEffect<ReposFailure>
  | PutEffect<ReposFulfill>,
  void,
  unknown
  >;

  let selectDescriptor: any;
  let callDescriptor: any;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getReposGenerator = getRepos({ type: 'boilerplate/Repos/REPOS_TRIGGER', payload: 'name' });

    selectDescriptor = getReposGenerator.next().value;

    callDescriptor = getReposGenerator.next().value;
  });

  it('should match the snapshot', () => {
    expect(selectDescriptor).toMatchSnapshot();

    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the reposLoaded action if it requests the data successfully', () => {
    const response = [
      {
        id: 1,
        name: 'First repo',
      },
      {
        id: 2,
        name: 'Second repo',
      },
    ];

    const putDescriptor = getReposGenerator.next(response).value;

    expect(putDescriptor).toEqual(put(reposSuccess(response)));
  });

  it('should call the repoLoadingError action if the response errors', () => {
    const response = new Error('Some error');

    const putDescriptor = getReposGenerator.throw(response).value;

    expect(putDescriptor).toEqual(put(reposFailure(response)));
  });
});

describe('reposDateSaga Saga', () => {
  const reposDateSaga = reposData();

  it('should start task to watch for REPOS_TRIGGER action', () => {
    const takeLatestDescriptor = reposDateSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(REPOS_TRIGGER, getRepos));
  });
});
