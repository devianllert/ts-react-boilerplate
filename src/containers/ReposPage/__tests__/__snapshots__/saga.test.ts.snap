// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getRepos Saga should match the snapshot 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "type": "boilerplate/Repos/REPOS_REQUEST",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`getRepos Saga should match the snapshot 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "name",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
